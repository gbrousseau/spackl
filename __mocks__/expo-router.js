const React = require('react');
const { View, Text } = require('react-native');
console.log('expo-router mock loaded');
module.exports = {
  useRouter: jest.fn(),
  useLocalSearchParams: jest.fn(),
  Redirect: ({ href }) => (
    React.createElement(View, null, React.createElement(Text, null, `Redirected to ${href}`))
  ),
};