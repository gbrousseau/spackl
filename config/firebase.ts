import firebaseAuth from '@react-native-firebase/auth';
import firebaseFirestore from '@react-native-firebase/firestore';
import { GoogleSignin } from '@react-native-google-signin/google-signin';

// Configure Google Sign-In
if (GoogleSignin) {
  GoogleSignin.configure({
    webClientId: process.env.EXPO_PUBLIC_GOOGLE_WEB_CLIENT_ID,
  });
}

export const auth = firebaseAuth;
export const firestore = firebaseFirestore;
export { GoogleSignin };
