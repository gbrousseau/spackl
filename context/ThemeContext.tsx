import React, { createContext, useContext, useState, useEffect } from 'react';
import { useColorScheme } from 'react-native';

interface ThemeContextType {
  isDark: boolean;
  toggleTheme: () => void;
  theme: 'light' | 'dark';
}

const ThemeContext = createContext<ThemeContextType>({
  isDark: false,
  toggleTheme: () => {},
  theme: 'light',
});

export const ThemeProvider: React.FC<{ children: React.ReactNode }> = ({
  children,
}) => {
  const colorScheme = useColorScheme();
  const [isDark, setIsDark] = useState(colorScheme === 'dark');
  const [theme, setTheme] = useState<'light' | 'dark'>(colorScheme || 'light');

  useEffect(() => {
    setIsDark(colorScheme === 'dark');
    setTheme(colorScheme || 'light');
  }, [colorScheme]);

  const toggleTheme = () => {
    setIsDark(!isDark);
    setTheme(isDark ? 'light' : 'dark');
  };

  return (
    <ThemeContext.Provider value={{ isDark, toggleTheme, theme }}>
      {children}
    </ThemeContext.Provider>
  );
};

export const useTheme = () => useContext(ThemeContext);
