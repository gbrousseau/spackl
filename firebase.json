{
  "apiKey": "AIzaSyCv0D1NJP4A-SeI1CS7W9TolAfefTwTunE",
  "authDomain": "spackl.firebaseapp.com",
  "projectId": "spackl",
  "storageBucket": "spackl.firebasestorage.app",
  "messagingSenderId": "913576495318",
  "appId": "1:913576495318:web:44aa95f69e3c5d6b2ce177"
}