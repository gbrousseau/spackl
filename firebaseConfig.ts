// Import the functions you need from the SDKs you need
import { initializeApp } from "firebase/app";
import { initializeAuth } from "firebase/auth";
import { getStorage } from "firebase/storage";
import { getFirestore } from "firebase/firestore";

const firebaseConfig = {
    apiKey: "AIzaSyBHaxJP1XKDc7iuY7ZfJAcBb7Jkgw7rsVw",
    authDomain: "spacklapp.firebaseapp.com",
    projectId: "spacklapp",
    storageBucket: "spacklapp.firebasestorage.app",
    messagingSenderId: "954071843248",
    appId: "1:954071843248:web:9b98c0fea6faa4480c5675",
    measurementId: "G-6F24MS1RKX"
};

// Initialize Firebase
export const FIREBASE_APP = initializeApp(firebaseConfig);
// Initialize Auth with AsyncStorage persistence
export const FIREBASE_AUTH = initializeAuth(FIREBASE_APP, {
  // persistence is not supported directly, remove or replace with a compatible library
});
export const FIREBASE_STORAGE = getStorage(FIREBASE_APP);
export const FIREBASE_FIRESTORE = getFirestore(FIREBASE_APP);