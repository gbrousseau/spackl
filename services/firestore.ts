import { FIREBASE_AUTH, FIREBASE_FIRESTORE } from '@/firebaseConfig';
import type { CalendarEvent, EventInvitation } from '@/types/calendar';
import type { DocumentData } from '@firebase/firestore-types';
import { getFirestore, collection, doc, setDoc, getDoc, updateDoc, serverTimestamp, query, where, getDocs, writeBatch } from 'firebase/firestore';

const auth = FIREBASE_AUTH;
const firestore = getFirestore();

export async function syncUserProfile(email: string, displayName: string) {
  try {
    await setDoc(doc(collection(firestore, 'users'), email), {
      email,
      displayName,
      updatedAt: serverTimestamp(),
    });
  } catch (err) {
    console.error('Failed to sync user profile:', err);
    throw err;
  }
}

export async function getUserProfile(email: string) {
  try {
    const docRef = doc(collection(firestore, 'users'), email);
    const docSnap = await getDoc(docRef);
    return docSnap.data() as DocumentData | undefined;
  } catch (err) {
    console.error('Failed to get user profile:', err);
    throw err;
  }
}

export async function syncCalendarEvent(event: CalendarEvent) {
  try {
    const user = auth.currentUser;
    if (!user?.email) throw new Error('User not authenticated');

    // Save event to Firestore
    await setDoc(doc(collection(firestore, 'users', user.email, 'calendar'), event.id), {
      ...event,
      updatedAt: serverTimestamp(),
    });

    // Create invitations for attendees
    if (event.attendees?.length) {
      const batch = writeBatch(firestore);

      for (const attendee of event.attendees) {
        if (attendee.email === user.email) continue;

        const invitationRef = doc(collection(firestore, 'users', attendee.email, 'invitations'));

        batch.set(invitationRef, {
          id: invitationRef.id,
          event,
          invitedBy: user.email,
          invitedEmail: attendee.email,
          status: 'pending',
          createdAt: serverTimestamp(),
        });
      }

      await batch.commit();
    }
  } catch (err) {
    console.error('Failed to sync calendar event:', err);
    throw err;
  }
}

export async function getCalendarEvents(email: string): Promise<CalendarEvent[]> {
  try {
    const q = query(collection(firestore, 'users', email, 'calendar'));
    const snapshot = await getDocs(q);

    return snapshot.docs.map((doc) => doc.data() as CalendarEvent);
  } catch (err) {
    console.error('Failed to get calendar events:', err);
    throw err;
  }
}

export async function getPendingInvitations(email: string): Promise<EventInvitation[]> {
  try {
    const q = query(collection(firestore, 'users', email, 'invitations'), where('status', '==', 'pending'));
    const snapshot = await getDocs(q);

    return snapshot.docs.map((doc) => {
      const data = doc.data();
      return {
        id: doc.id,
        event: data.event,
        invitedBy: data.invitedBy,
        invitedEmail: data.invitedEmail,
        status: data.status,
        createdAt: data.createdAt.toDate(),
      } as EventInvitation;
    });
  } catch (err) {
    console.error('Failed to get pending invitations:', err);
    throw err;
  }
}

export async function respondToInvitation(invitationId: string, email: string, status: 'accepted' | 'declined') {
  try {
    const invitationRef = doc(collection(firestore, 'users', email, 'invitations'), invitationId);
    const invitationDoc = await getDoc(invitationRef);
    if (!invitationDoc.exists()) throw new Error('Invitation not found');

    const invitation = invitationDoc.data() as EventInvitation;

    // Update invitation status
    await updateDoc(invitationRef, {
      status,
      updatedAt: serverTimestamp(),
    });

    // If accepted, add event to user's calendar
    if (status === 'accepted') {
      await setDoc(doc(collection(firestore, 'users', email, 'calendar'), invitation.event.id), {
        ...invitation.event,
        updatedAt: serverTimestamp(),
      });
    }
  } catch (err) {
    console.error('Failed to respond to invitation:', err);
    throw err;
  }
}
