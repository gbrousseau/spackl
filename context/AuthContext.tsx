import { createContext, useContext, useState, useEffect } from 'react';
import { useRouter, useSegments } from 'expo-router';
import {
  User,
  signInWithEmailAndPassword,
  createUserWithEmailAndPassword,
  signOut as firebaseSignOut,
  onAuthStateChanged,
  GoogleAuthProvider,
  signInWithCredential,
} from 'firebase/auth';
import { FIREBASE_AUTH } from '@/firebaseConfig';
import { doc, getDoc, setDoc, updateDoc, collection, query, where, getDocs } from 'firebase/firestore';
import { FIREBASE_FIRESTORE } from '@/firebaseConfig';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { Platform, PermissionsAndroid } from 'react-native';
import * as Device from 'expo-device';

type AuthContextType = {
  user: User | null;
  deviceInfo: {
    platform: string;
    model: string;
  };
  phoneInfo: {
    phoneNumber: string;
    countryCode: string;
  } | null;
  signUp: (email: string, password: string) => Promise<void>;
  signIn: (email: string, password: string) => Promise<void>;
  signInWithGoogle: (idToken: string) => Promise<void>;
  signOut: () => Promise<void>;
  loading: boolean;
  error: string | null;
  clearError: () => void;
  requestPermissions: () => Promise<boolean>;
};

export const AuthContext = createContext<AuthContextType>({
  user: null,
  deviceInfo: {
    platform: Platform.OS,
    model: 'Unknown Device'
  },
  phoneInfo: null,
  signUp: async () => {},
  signIn: async () => {},
  signInWithGoogle: async () => {},
  signOut: async () => {},
  loading: true,
  error: null,
  clearError: () => {},
  requestPermissions: async () => false,
});

// This hook will protect the route access based on user authentication
function useProtectedRoute(user: User | null, isLoading: boolean) {
  const segments = useSegments();
  const router = useRouter();

  useEffect(() => {
    if (isLoading) return;

    const inAuthGroup = segments[0] === 'auth';

    if (!user && !inAuthGroup) {
      // Redirect to the sign-in page if not authenticated
      router.replace('/auth/login');
    } else if (user && inAuthGroup) {
      // Redirect away from the sign-in page if authenticated
      router.replace('/(tabs)');
    }
  }, [user, isLoading, segments]);
}

export function AuthProvider({ children }: { children: React.ReactNode }) {
  const [user, setUser] = useState<User | null>(null);
  const [deviceInfo, setDeviceInfo] = useState({
    platform: Platform.OS,
    model: 'Unknown Device'
  });
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState<string | null>(null);
  const [phoneInfo, setPhoneInfo] = useState<{
    phoneNumber: string;
    countryCode: string;
  } | null>(null);
  const router = useRouter();
  
  useProtectedRoute(user, loading);

  const requestPermissions = async (): Promise<boolean> => {
    if (Platform.OS === 'web') {
      return true;
    }

    try {
      if (Platform.OS === 'android') {
        const granted = await PermissionsAndroid.requestMultiple([
          PermissionsAndroid.PERMISSIONS.READ_PHONE_STATE,
          PermissionsAndroid.PERMISSIONS.READ_CONTACTS,
        ]);

        return Object.values(granted).every(
          permission => permission === PermissionsAndroid.RESULTS.GRANTED
        );
      }

      // For iOS, permissions are handled through Info.plist
      return true;
    } catch (err) {
      console.error('Error requesting permissions:', err);
      return false;
    }
  };

  useEffect(() => {
    const getDeviceInfo = async () => {
      try {
        const hasPermissions = await requestPermissions();
        if (!hasPermissions) {
          console.warn('Required permissions not granted');
          return;
        }

        const model = await Device.modelId || await Device.deviceName || 'Unknown Device';
        setDeviceInfo({
          platform: Platform.OS,
          model
        });
      } catch (error) {
        console.error('Error getting device info:', error);
      }
    };
    
    getDeviceInfo();
  }, []);

  useEffect(() => {
    const unsubscribe = onAuthStateChanged(FIREBASE_AUTH, async (user) => {
      setUser(user);
      setLoading(false);

      if (user) {
        // Store user session
        await AsyncStorage.setItem('user', JSON.stringify(user));
        await AsyncStorage.setItem('deviceInfo', JSON.stringify(deviceInfo));

        // Check if user exists in users collection
        const userRef = doc(FIREBASE_FIRESTORE, 'users', user.uid);
        const userDoc = await getDoc(userRef);

        // Get user's groups
        const groupsQuery = query(
          collection(FIREBASE_FIRESTORE, 'groups'),
          where('members', 'array-contains', user.uid)
        );
        const groupsSnapshot = await getDocs(groupsQuery);
        const userGroups = groupsSnapshot.docs.map(doc => ({
          id: doc.id,
          ...doc.data()
        }));

        if (!userDoc.exists()) {
          // Create new user document
          await setDoc(userRef, {
            email: user.email,
            displayName: user.displayName || '',
            phoneNumber: user.phoneNumber || '',
            photoURL: user.photoURL || '',
            createdAt: new Date(),
            lastLogin: new Date(),
            settings: {
              notifications: true,
              darkMode: false,
            },
            groups: userGroups,
            deviceInfo
          });
        } else {
          // Update existing user document
          await updateDoc(userRef, {
            email: user.email,
            displayName: user.displayName || '',
            phoneNumber: user.phoneNumber || '',
            photoURL: user.photoURL || '',
            lastLogin: new Date(),
            groups: userGroups,
            deviceInfo
          });
        }
      } else {
        await AsyncStorage.removeItem('user');
        await AsyncStorage.removeItem('deviceInfo');
      }
    });

    // Check for stored session on mount
    AsyncStorage.getItem('user').then((storedUser) => {
      if (storedUser) {
        setUser(JSON.parse(storedUser));
      }
      setLoading(false);
    });

    return () => unsubscribe();
  }, [deviceInfo]);

  const signUp = async (email: string, password: string) => {
    try {
      setError(null);
      setLoading(true);

      const hasPermissions = await requestPermissions();
      if (!hasPermissions) {
        throw new Error('Required permissions not granted');
      }

      const deviceInfo = {
        modelName: await Device.modelName || 'unknown',
        deviceName: await Device.deviceName || 'unknown',
        brand: await Device.brand || 'unknown',
        manufacturer: await Device.manufacturer || 'unknown',
        osName: await Device.osName || 'unknown',
        osVersion: await Device.osVersion || 'unknown',
        platformApiLevel: await Device.platformApiLevel || 'unknown'
      };

      const userCredential = await createUserWithEmailAndPassword(FIREBASE_AUTH, email, password);
      const user = userCredential.user;

      // Create user document in Firestore
      await setDoc(doc(FIREBASE_FIRESTORE, 'users', user.uid), {
        email: user.email,
        displayName: user.displayName || '',
        phoneNumber: deviceInfo.modelName,
        photoURL: user.photoURL || '',
        createdAt: new Date(),
        lastLogin: new Date(),
        settings: {
          notifications: true,
          darkMode: false,
        },
        groups: [],
        deviceInfo
      });

      setUser(user);
      router.replace('/(tabs)');
    } catch (error: any) {
      setError(error.message);
      throw error;
    } finally {
      setLoading(false);
    }
  };

  const signIn = async (email: string, password: string) => {
    try {
      setError(null);
      setLoading(true);

      const hasPermissions = await requestPermissions();
      if (!hasPermissions) {
        throw new Error('Required permissions not granted');
      }

      const deviceInfo = {
        modelName: await Device.modelName || 'unknown',
        deviceName: await Device.deviceName || 'unknown',
        brand: await Device.brand || 'unknown',
        manufacturer: await Device.manufacturer || 'unknown',
        osName: await Device.osName || 'unknown',
        osVersion: await Device.osVersion || 'unknown',
        platformApiLevel: await Device.platformApiLevel || 'unknown'
      };

      const userCredential = await signInWithEmailAndPassword(FIREBASE_AUTH, email, password);
      const user = userCredential.user;

      // Update user document in Firestore
      await updateDoc(doc(FIREBASE_FIRESTORE, 'users', user.uid), {
        lastLogin: new Date(),
        deviceInfo
      });

      setUser(user);
      router.replace('/(tabs)');
    } catch (error: any) {
      setError(error.message);
      throw error;
    } finally {
      setLoading(false);
    }
  };

  const signInWithGoogle = async (idToken: string) => {
    try {
      setError(null);
      setLoading(true);

      const hasPermissions = await requestPermissions();
      if (!hasPermissions) {
        throw new Error('Required permissions not granted');
      }

      const credential = GoogleAuthProvider.credential(idToken);
      const userCredential = await signInWithCredential(FIREBASE_AUTH, credential);
      setUser(userCredential.user);
      router.replace('/(tabs)');
    } catch (error: any) {
      setError(error.message);
      throw error;
    } finally {
      setLoading(false);
    }
  };

  const signOut = async () => {
    try {
      setError(null);
      setLoading(true);
      await firebaseSignOut(FIREBASE_AUTH);
      setUser(null);
      await AsyncStorage.removeItem('user');
      router.replace('/auth/login');
    } catch (error: any) {
      setError(error.message);
      throw error;
    } finally {
      setLoading(false);
    }
  };

  const getPhoneInfo = async () => {
    try {
      if (Platform.OS === 'web') {
        setPhoneInfo({
          phoneNumber: '1234567890',
          countryCode: '+1'
        });
        return;
      }

      const hasPermissions = await requestPermissions();
      if (!hasPermissions) {
        console.warn('Required permissions not granted for phone info');
        return;
      }

      const deviceName = await Device.deviceName;
      const modelName = await Device.modelName;
      
      // Use a combination of device name and model as a unique identifier
      const deviceIdentifier = `${deviceName}-${modelName}`;
      
      if (deviceIdentifier) {
        setPhoneInfo({
          phoneNumber: deviceIdentifier,
          countryCode: '+1' // Default to +1 since we can't reliably get country code
        });
      }
    } catch (err) {
      console.error('Error getting phone info:', err);
      setPhoneInfo({
        phoneNumber: 'unknown',
        countryCode: '+1'
      });
    }
  };

  useEffect(() => {
    if (user) {
      getPhoneInfo();
    } else {
      setPhoneInfo(null);
    }
  }, [user]);

  return (
    <AuthContext.Provider
      value={{
        user,
        deviceInfo,
        phoneInfo,
        signUp,
        signIn,
        signInWithGoogle,
        signOut,
        loading,
        error,
        clearError: () => setError(null),
        requestPermissions,
      }}>
      {children}
    </AuthContext.Provider>
  );
}

export const useAuth = () => useContext(AuthContext);