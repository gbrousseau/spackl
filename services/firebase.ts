import { initializeApp } from 'firebase/app';

// Define CalendarEvent type
export interface CalendarEvent {
  id: string;
  title: string;
  start: string;
  end: string;
  location?: string;
  description?: string;
}
import { getFirestore, doc, setDoc, getDoc } from 'firebase/firestore';
import { getAuth } from 'firebase/auth';
import AsyncStorage from '@react-native-async-storage/async-storage';

const firebaseConfig = {
  apiKey: process.env.EXPO_PUBLIC_FIREBASE_API_KEY,
  authDomain: process.env.EXPO_PUBLIC_FIREBASE_AUTH_DOMAIN,
  projectId: process.env.EXPO_PUBLIC_FIREBASE_PROJECT_ID,
  storageBucket: process.env.EXPO_PUBLIC_FIREBASE_STORAGE_BUCKET,
    // events: CalendarEvent[]; // Removed invalid line
  appId: process.env.EXPO_PUBLIC_FIREBASE_APP_ID,
};

const app = initializeApp(firebaseConfig);
const db = getFirestore(app);
const auth = getAuth(app);

export interface AppState {
  calendar: {
    events: CalendarEvent[];
    lastSync: string;
  };
  settings: {
    theme: 'light' | 'dark' | 'system';
    notifications: boolean;
    defaultCalendar?: string;
    defaultView: 'month' | 'week' | 'day';
    [key: string]: any;
  };
  [key: string]: any;
}

export const syncWithFirebase = async (
  userId: string,
  state: Partial<AppState>,
) => {
  try {
    const userDoc = doc(db, 'users', userId);
    await setDoc(userDoc, state, { merge: true });
    await AsyncStorage.setItem(`appState_${userId}`, JSON.stringify(state));
    return true;
  } catch (error) {
    console.error('Error syncing with Firebase:', error);
    return false;
  }
};

export const loadFromFirebase = async (
  userId: string,
): Promise<Partial<AppState> | null> => {
  try {
    const userDoc = doc(db, 'users', userId);
    const docSnap = await getDoc(userDoc);

    if (docSnap.exists()) {
      const data = docSnap.data() as Partial<AppState>;
      await AsyncStorage.setItem(`appState_${userId}`, JSON.stringify(data));
      return data;
    }
    return null;
  } catch (error) {
    console.error('Error loading from Firebase:', error);
    return null;
  }
};

export const loadFromLocal = async (
  userId: string,
): Promise<Partial<AppState> | null> => {
  try {
    const localData = await AsyncStorage.getItem(`appState_${userId}`);
    if (localData) {
      return JSON.parse(localData);
    }
    return null;
  } catch (error) {
    console.error('Error loading from local storage:', error);
    return null;
  }
};

export const syncCalendarEvents = async (
  userId: string,
  events: CalendarEvent[],
) => {
  const state: Partial<AppState> = {
    calendar: {
      events,
      lastSync: new Date().toISOString(),
    },
  };
  return syncWithFirebase(userId, state);
};

export const syncAppSettings = async (
  userId: string,
  settings: AppState['settings'],
) => {
  const state: Partial<AppState> = {
    settings,
  };
  return syncWithFirebase(userId, state);
};

export { db, auth };
