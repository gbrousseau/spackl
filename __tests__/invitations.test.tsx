import { render, screen, fireEvent, waitFor } from '@testing-library/react-native';
import { useRouter } from 'expo-router';
import { collection, getDocs } from 'firebase/firestore';
import { FIREBASE_FIRESTORE } from '@/config/firebaseConfig';
import { useAuth } from '@/contexts/AuthContext';
import { useTheme } from '@/contexts/ThemeContext';
import InvitationsScreen from '@/app/(tabs)/invitations';
import * as ExpoCalendar from 'expo-calendar';

// Mock expo-router
jest.mock('expo-router', () => ({
  useRouter: jest.fn(),
}));

// Mock firebase/firestore
jest.mock('firebase/firestore', () => ({
  collection: jest.fn(),
  getDocs: jest.fn(),
}));

// Mock contexts
jest.mock('../contexts/AuthContext', () => ({
  useAuth: jest.fn(),
}));

jest.mock('../contexts/ThemeContext', () => ({
  useTheme: jest.fn(),
}));

// Mock expo-calendar
jest.mock('expo-calendar', () => ({
  requestCalendarPermissionsAsync: jest.fn(),
  getCalendarsAsync: jest.fn(),
  createEventAsync: jest.fn(),
}));

// Mock data
const mockUser = {
  phoneNumber: '+1234567890',
  email: 'test@example.com',
};

const mockInvitations = [
  {
    id: '1',
    eventId: 'event1',
    status: 'pending',
    event: {
      title: 'Test Event 1',
      description: 'Test Description 1',
      startTime: new Date().toISOString(),
      endTime: new Date().toISOString(),
      location: 'Test Location 1',
      organizer: {
        name: 'Test Organizer 1',
      },
    },
  },
  {
    id: '2',
    eventId: 'event2',
    status: 'going',
    event: {
      title: 'Test Event 2',
      description: 'Test Description 2',
      startTime: new Date().toISOString(),
      endTime: new Date().toISOString(),
      location: 'Test Location 2',
      organizer: {
        name: 'Test Organizer 2',
      },
    },
  },
  {
    id: '3',
    eventId: 'event3',
    status: 'interested',
    event: {
      title: 'Test Event 3',
      description: 'Test Description 3',
      startTime: new Date().toISOString(),
      endTime: new Date().toISOString(),
      location: 'Test Location 3',
      organizer: {
        name: 'Test Organizer 3',
      },
    },
  },
];

describe('InvitationsScreen', () => {
  beforeEach(() => {
    // Reset all mocks
    jest.clearAllMocks();

    // Setup default mock implementations
    (useRouter as jest.Mock).mockReturnValue({ push: jest.fn() });
    (useAuth as jest.Mock).mockReturnValue({ user: mockUser });
    (useTheme as jest.Mock).mockReturnValue({ isDarkMode: false });

    // Setup Firestore mocks
    (collection as jest.Mock).mockImplementation((db, ...path) => ({
      path: path.join('/'),
    }));

    (getDocs as jest.Mock).mockImplementation(async (ref) => ({
      docs: mockInvitations.map(invitation => ({
        id: invitation.id,
        data: () => invitation,
      })),
    }));
  });

  it('renders loading state initially', async () => {
    render(<InvitationsScreen />);
    
    // Check for loading indicator
    expect(screen.getByTestId('loading-indicator')).toBeTruthy();
    
    // Wait for loading to complete
    await waitFor(() => {
      expect(screen.queryByTestId('loading-indicator')).toBeNull();
    });
  });

  it('displays invitations grouped by status', async () => {
    render(<InvitationsScreen />);

    // Wait for loading to complete
    await waitFor(() => {
      // Check for pending invitations
      expect(screen.getByText('Pending')).toBeTruthy();
      expect(screen.getByText('Test Event 1')).toBeTruthy();
      expect(screen.getByText('Test Organizer 1')).toBeTruthy();

      // Check for going invitations
      expect(screen.getByText('Going')).toBeTruthy();
      expect(screen.getByText('Test Event 2')).toBeTruthy();
      expect(screen.getByText('Test Organizer 2')).toBeTruthy();

      // Check for interested invitations
      expect(screen.getByText('Interested')).toBeTruthy();
      expect(screen.getByText('Test Event 3')).toBeTruthy();
      expect(screen.getByText('Test Organizer 3')).toBeTruthy();
    });
  });

  it('handles invitation click', async () => {
    const mockPush = jest.fn();
    (useRouter as jest.Mock).mockReturnValue({ push: mockPush });

    render(<InvitationsScreen />);

    // Wait for loading to complete
    await waitFor(() => {
      expect(screen.getByText('Test Event 1')).toBeTruthy();
    });

    // Click on an invitation
    fireEvent.press(screen.getByText('Test Event 1'));

    // Verify router was called with correct path
    expect(mockPush).toHaveBeenCalledWith('/invitation/1');
  });

  it('handles error state when loading fails', async () => {
    // Mock getDocs to throw an error
    (getDocs as jest.Mock).mockRejectedValueOnce(new Error('Failed to load invitations'));

    render(<InvitationsScreen />);

    // Wait for error message
    await waitFor(() => {
      expect(screen.getByText('Failed to load invitations')).toBeTruthy();
    });

    // Check for retry button
    expect(screen.getByText('Retry')).toBeTruthy();
  });

  it('attempts multiple identifiers when loading invitations', async () => {
    render(<InvitationsScreen />);

    // Wait for loading to complete
    await waitFor(() => {
      expect(screen.getByText('Test Event 1')).toBeTruthy();
    });

    // Verify collection was called with both phone number and email
    expect(collection).toHaveBeenCalledWith(
      FIREBASE_FIRESTORE,
      'users',
      mockUser.phoneNumber,
      'invitations'
    );
    expect(collection).toHaveBeenCalledWith(
      FIREBASE_FIRESTORE,
      'users',
      mockUser.email,
      'invitations'
    );
  });

  it('displays empty state when no invitations exist', async () => {
    // Mock getDocs to return empty array
    (getDocs as jest.Mock).mockImplementationOnce(async () => ({
      docs: [],
    }));

    render(<InvitationsScreen />);

    // Wait for loading to complete
    await waitFor(() => {
      expect(screen.queryByTestId('loading-indicator')).toBeNull();
    });

    // Check for empty state message
    expect(screen.getByText('You don\'t have any event invitations yet')).toBeTruthy();
    expect(screen.getByTestId('empty-state-icon')).toBeTruthy();
  });

  it('handles calendar permission denied', async () => {
    // Mock calendar permission to be denied
    (ExpoCalendar.requestCalendarPermissionsAsync as jest.Mock).mockResolvedValueOnce({
      status: 'denied',
    });

    render(<InvitationsScreen />);

    // Wait for loading to complete
    await waitFor(() => {
      expect(screen.getByText('Test Event 1')).toBeTruthy();
    });

    // Click on an invitation
    fireEvent.press(screen.getByText('Test Event 1'));

    // Check for permission alert
    await waitFor(() => {
      expect(screen.getByText('Calendar Permission Required')).toBeTruthy();
    });
  });

  it('adds event to calendar when status is set to going', async () => {
    // Mock calendar permission to be granted
    (ExpoCalendar.requestCalendarPermissionsAsync as jest.Mock).mockResolvedValueOnce({
      status: 'granted',
    });

    // Mock getCalendarsAsync to return a default calendar
    (ExpoCalendar.getCalendarsAsync as jest.Mock).mockResolvedValueOnce([
      { id: 'default-calendar', isPrimary: true },
    ]);

    render(<InvitationsScreen />);

    // Wait for loading to complete
    await waitFor(() => {
      expect(screen.getByText('Test Event 1')).toBeTruthy();
    });

    // Click on an invitation
    fireEvent.press(screen.getByText('Test Event 1'));

    // Wait for calendar event to be created
    await waitFor(() => {
      expect(ExpoCalendar.createEventAsync).toHaveBeenCalledWith(
        'default-calendar',
        expect.objectContaining({
          title: 'Test Event 1',
          startDate: expect.any(Date),
          endDate: expect.any(Date),
        })
      );
    });
  });

  it('sorts invitations by date (newest first)', async () => {
    const mockInvitationsWithDates = [
      {
        id: '1',
        eventId: 'event1',
        status: 'pending',
        event: {
          title: 'Older Event',
          startTime: new Date('2023-01-01').toISOString(),
          endTime: new Date('2023-01-02').toISOString(),
        },
      },
      {
        id: '2',
        eventId: 'event2',
        status: 'pending',
        event: {
          title: 'Newer Event',
          startTime: new Date('2023-02-01').toISOString(),
          endTime: new Date('2023-02-02').toISOString(),
        },
      },
    ];

    (getDocs as jest.Mock).mockImplementationOnce(async () => ({
      docs: mockInvitationsWithDates.map(invitation => ({
        id: invitation.id,
        data: () => invitation,
      })),
    }));

    render(<InvitationsScreen />);

    // Wait for loading to complete
    await waitFor(() => {
      const invitations = screen.getAllByTestId('invitation-card');
      expect(invitations[0].props.children[0].props.children).toBe('Newer Event');
      expect(invitations[1].props.children[0].props.children).toBe('Older Event');
    });
  });

  it('handles dark mode styling', async () => {
    // Mock dark mode
    (useTheme as jest.Mock).mockReturnValueOnce({ isDarkMode: true });

    render(<InvitationsScreen />);

    // Wait for loading to complete
    await waitFor(() => {
      expect(screen.getByText('Test Event 1')).toBeTruthy();
    });

    // Check for dark mode styles
    const invitationCard = screen.getByTestId('invitation-card');
    expect(invitationCard.props.style).toContainEqual(
      expect.objectContaining({
        backgroundColor: expect.stringMatching(/dark/i),
      })
    );
  });

  it('handles network error when loading invitations', async () => {
    // Mock getDocs to throw a network error
    (getDocs as jest.Mock).mockRejectedValueOnce(new Error('Network error'));

    render(<InvitationsScreen />);

    // Wait for error message
    await waitFor(() => {
      expect(screen.getByText('Network error')).toBeTruthy();
    });

    // Check for retry button
    expect(screen.getByText('Retry')).toBeTruthy();
  });

  it('updates invitation status and adds to calendar', async () => {
    // Mock calendar permission to be granted
    (ExpoCalendar.requestCalendarPermissionsAsync as jest.Mock).mockResolvedValueOnce({
      status: 'granted',
    });

    // Mock getCalendarsAsync to return a default calendar
    (ExpoCalendar.getCalendarsAsync as jest.Mock).mockResolvedValueOnce([
      { id: 'default-calendar', isPrimary: true },
    ]);

    render(<InvitationsScreen />);

    // Wait for loading to complete
    await waitFor(() => {
      expect(screen.getByText('Test Event 1')).toBeTruthy();
    });

    // Click on an invitation
    fireEvent.press(screen.getByText('Test Event 1'));

    // Wait for calendar event to be created
    await waitFor(() => {
      expect(ExpoCalendar.createEventAsync).toHaveBeenCalledWith(
        'default-calendar',
        expect.objectContaining({
          title: 'Test Event 1',
          startDate: expect.any(Date),
          endDate: expect.any(Date),
        })
      );
    });

    // Check for success message
    expect(screen.getByText('Event added to your calendar')).toBeTruthy();
  });

  it('handles calendar permission denied gracefully', async () => {
    // Mock calendar permission to be denied
    (ExpoCalendar.requestCalendarPermissionsAsync as jest.Mock).mockResolvedValueOnce({
      status: 'denied',
    });

    render(<InvitationsScreen />);

    // Wait for loading to complete
    await waitFor(() => {
      expect(screen.getByText('Test Event 1')).toBeTruthy();
    });

    // Click on an invitation
    fireEvent.press(screen.getByText('Test Event 1'));

    // Check for permission alert
    await waitFor(() => {
      expect(screen.getByText('Calendar Permission Required')).toBeTruthy();
    });

    // Check that calendar event was not created
    expect(ExpoCalendar.createEventAsync).not.toHaveBeenCalled();
  });
}); 