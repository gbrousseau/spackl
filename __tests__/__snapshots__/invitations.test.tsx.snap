// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InvitationsScreen expo-router mock should use the mocked Redirect 1`] = `"Redirected to /test"`;
